Modify GPU Resource Declaration When Not Requested

namrathaurs opened the pull request:
Fixes #444.

Discussed with Marco to understand the existing logic and behavior of using `GLIDEIN_Resource_Slots` attribute in factory configuration. The details are as follows:

To cater to the new behavior change in HTCondor when it comes to GPUs, three conditions were necessary to be included in order to set GPUS explicitly to 0 when GPUs are not requested. The three conditions are:
1. when `GLIDEIN_Resource_Slots` attribute is not defined in factory configuration, or
2. when `GLIDEIN_Resource_Slots` attribute's value does not include `GPUs`, or
3. when `GLIDEIN_Resource_Slots` attribute's value is set to 0.

**Logic to cater to behavior arising from conditions (1) and (2) is added newly as part of this PR while previously existing logic already had the behavior defined when condition (3) occurs.**

<hr />


**Additional details:**
Following examples of `GLIDEIN_Resource_Slots` attribute were used for testing all of the three conditions which constitute the updated logic: 
```
<!-- for testing condition (2) -->
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="ioslot,2,200,static"/>
```
```
<!-- for testing condition (3) -->
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="GPUs,0,type=main"/>
```

Following were also used for generic testing and debugging of the existing code before the changes were made to `condor_startup.sh`:
```
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="True" job_publish="False" parameter="True" publish="True" type="string" value="GPUs,1,type=main"/>
```
```
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="True" job_publish="False" parameter="True" publish="True" type="string" value="ioslot,2,disk=1GB;monitor;GPUs,3,,main"/>
```

namrathaurs commented:
Adding another test:
GPUs is set to AUTO and document the results of this test. it might say zero and it's okay

GPUs set to a value other than zero 

And the previous tests

mambelli commented on creation/web_base/condor_startup.sh:
```diff
@@ -880,6 +880,9 @@ EOF
                     cat >> "$CONDOR_CONFIG" <<EOF
 # Declare resource: ${i}
 MACHINE_RESOURCE_${res_name} = ${res_num}
+# Declare no use of GPUs, i.e. the worker node (WN) should not have GPUs.
+# Even though the WN may physically have them, we want them to be ignored.
+MACHINE_RESOURCE_GPUs = 0
```
This may run in the loop and be printed multiple times, even if GPUs are present

namrathaurs replied:
This ended up being the case and has been modified as needed.

namrathaurs commented:
Test cases (TCs) and their outputs:

- **TC0:-** when `GLIDEIN_Resource_Slots` undefined in factory config: found `MACHINE_RESOURCE_GPUs = 0` in condor_config written
- **TC1:-**  when `GLIDEIN_Resource_Slots` attribute's value does not include `GPUs` ⇒ i.e. `... value="ioslot,2,200,static"` : found `MACHINE_RESOURCE_GPUs = 0` in condor_config written
- **TC2:-** when `GLIDEIN_Resource_Slots` attribute’s value does not include GPUs ⇒ i.e. `... "value= ioslot,2,200,static;ioslot1,1,128,static"` : found `MACHINE_RESOURCE_GPUs = 0` in condor_config written along with `MACHINE_RESOURCE_ioslot = 2` and `MACHINE_RESOURCE_ioslot1 = 1`
- **TC3:-** when `GLIDEIN_Resource_Slots` attribute’s value includes GPUs and is set to zero ⇒ i.e. `... value="GPUs,0,type=main"` : found `MACHINE_RESOURCE_GPUs = 0` in condor_config written
- **TC4:-** when `GLIDEIN_Resource_Slots` attribute’s value includes GPUs and is set to zero ⇒ i.e. `... value="ioslot,2,200,static;GPUs,0,type=main;ioslot1,1,128,static"` : found `MACHINE_RESOURCE_GPUs = 0` in condor_config written along with `MACHINE_RESOURCE_ioslot = 2` and `MACHINE_RESOURCE_ioslot1 = 1`
- **TC5:-** when `GLIDEIN_Resource_Slots` attribute’s value includes GPUs and is set to non-zero ⇒ i.e. `... value="GPUs,2,type=main"` : found `MACHINE_RESOURCE_GPUs = 2` in condor_config written
- **TC6:-** when `GLIDEIN_Resource_Slots` attribute’s value includes GPUs and is set to non-zero ⇒ i.e. `... value="ioslot,2,200,static;GPUs,2,type=main"` : found `MACHINE_RESOURCE_GPUs = 2` in condor_config written along with `MACHINE_RESOURCE_ioslot = 2`
- **TC7:-** when `GLIDEIN_Resource_Slots` attribute’s value includes GPUs and is set to non-zero ⇒ i.e. `... value="ioslot,2,disk=1GB;monitor;GPUs,3,main"` : found `MACHINE_RESOURCE_GPUs = 3` in condor_config written along with `MACHINE_RESOURCE_ioslot = 2` and `MACHINE_RESOURCE_monitor = 1`. ***However, it was observed that frontend jobs do not complete and therefore, more number of glideins get requested in the factory.***
- **TC8:-** when `GLIDEIN_Resource_Slots` attribute’s value includes GPUs and is set to AUTO ⇒ i.e. `... value="GPUs,AUTO,type=main"` : found `MACHINE_RESOURCE_GPUs = AUTO` in condor_config written

namrathaurs commented on creation/web_base/condor_startup.sh:
```diff
@@ -946,7 +949,13 @@ if defined MACHINE_RESOURCE_NAMES
 START = (\$(START)) && (\$(EXTRA_SLOTS_START))
 EOF
-
+        else
+	    # when GLIDEIN_Resource_Slots is not defined
+            cat >> "$CONDOR_CONFIG" <<EOF
+# Declare no use of GPUs, i.e. the worker node (WN) should not have GPUs.
+# Even though the WN may physically have them, we want them to be ignored.
+MACHINE_RESOURCE_GPUs = 0
+EOF
```
updated

namrathaurs commented:
> TC7:- when GLIDEIN_Resource_Slots attribute’s value includes GPUs and is set to non-zero ⇒ i.e. ... value="ioslot,2,disk=1GB;monitor;GPUs,3,main" : found MACHINE_RESOURCE_GPUs = 3 in condor_config written along with MACHINE_RESOURCE_ioslot = 2 and MACHINE_RESOURCE_monitor = 1. 
However, it was observed that frontend jobs do not complete and therefore, more number of glideins continue to be requested in the factory.

**UPDATE:** Upon further inspection, the frontend jobs were not completing because the glidein startup was not completing successfully (glidein stdout reported `
=== Condor started in background but the pid file is still missing, waiting 200 sec more ===`. This was something familiar as it was encountered previously where the MasterLog and StartdLog in the glidein would contain more details about the exact problem that is causing condor_startup to fail. Upon looking into the StartdLog, noticed the following:
```
10/30/24 10:32:00 (pid:3700140) You must specify a percent or fraction for disk in slot type 3
10/30/24 10:32:00 (pid:3700140) **** condor_startd (condor_STARTD) pid 3700140 EXITING WITH STATUS 4
```

The error message points to the factory configuration setting for `GLIDEIN_Resource_Slots` which was defined as:
```
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="ioslot,2,disk=1GB;monitor;GPUs,3,main"/>
```

Following the pointer in the error message, updated the setting to:
```
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="ioslot,2,disk=1GB;monitor;GPUs,3,main,10%"/>
```

namrathaurs commented:
**UPDATE:** Based on the above observations, Marco gave a suggestion to check the resulting `condor_config` in the glidein to see if that could help provide clues as to what could be going wrong. The resulting condor config when using 
```
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="ioslot,2,disk=1GB;monitor;GPUs,3,main"/>
```
is as follows:
```
...
# ---- start of include part ----
NUM_CPUS = $(GLIDEIN_CPUS)
SLOT_TYPE_1 = cpus=1
NUM_SLOTS_TYPE_1 = $(GLIDEIN_CPUS)
# ---- start of resource slots part (ioslot,2,disk=1GB;monitor;GPUs,3,main) ----
NEW_RESOURCES_LIST =
EXTRA_SLOTS_NUM = 0
EXTRA_CPUS_NUM = 0
EXTRA_SLOTS_START = True
NUM_CPUS = $(GLIDEIN_CPUS)+$(EXTRA_SLOTS_NUM)+$(EXTRA_CPUS_NUM)

# Slot 1 definition done before (fixed/partitionable)
#SLOT_TYPE_1_PARTITIONABLE = FALSE
#SLOT_TYPE_1 = cpus=1, ioslot=0
#NUM_SLOTS_TYPE_1 = $(GLIDEIN_CPUS)
#
#SLOT_TYPE_1_PARTITIONABLE = TRUE
#SLOT_TYPE_1 = ioslot=0
#NUM_SLOTS_TYPE_1 = 1
# Declare resource: ioslot,2,disk=1GB
MACHINE_RESOURCE_ioslot = 2
EXTRA_SLOTS_NUM = $(EXTRA_SLOTS_NUM)+$(MACHINE_RESOURCE_ioslot)
SLOT_TYPE_3 = cpus=$(MACHINE_RESOURCE_ioslot), ioslot=$(MACHINE_RESOURCE_ioslot), ram=256, disk=1GB
SLOT_TYPE_3_PARTITIONABLE = TRUE
NUM_SLOTS_TYPE_3 = 1
IS_SLOT_ioslot = SlotTypeID==3
EXTRA_SLOTS_START = ifThenElse((SlotTypeID==3), TARGET.Requestioslot>0, ($(EXTRA_SLOTS_START)))
NEW_RESOURCES_LIST = $(NEW_RESOURCES_LIST) ioslot
# Declare resource: monitor
MACHINE_RESOURCE_monitor = 1
EXTRA_SLOTS_NUM = $(EXTRA_SLOTS_NUM)+$(MACHINE_RESOURCE_monitor)
SLOT_TYPE_4 = cpus=1, monitor=1, ram=128
SLOT_TYPE_4_PARTITIONABLE = FALSE
NUM_SLOTS_TYPE_4 = $(MACHINE_RESOURCE_monitor)
IS_SLOT_monitor = SlotTypeID==4
EXTRA_SLOTS_START = ifThenElse((SlotTypeID==4), TARGET.Requestmonitor>0, ($(EXTRA_SLOTS_START)))
NEW_RESOURCES_LIST = $(NEW_RESOURCES_LIST) monitor
# Declare GPU resource, forcing 3: GPUs,3,main
use feature : GPUs
# GPUsMonitor is automatically included in newer HTCondor
use feature : GPUsMonitor
GPU_DISCOVERY_EXTRA = -extra
MACHINE_RESOURCE_GPUs = 3
EXTRA_SLOTS_NUM = $(EXTRA_SLOTS_NUM)+$(MACHINE_RESOURCE_GPUs)
SLOT_TYPE_5 = cpus=$(MACHINE_RESOURCE_GPUs), GPUs=$(MACHINE_RESOURCE_GPUs), ram=main
SLOT_TYPE_5_PARTITIONABLE = TRUE
NUM_SLOTS_TYPE_5 = 1
IS_SLOT_GPUs = SlotTypeID==5
EXTRA_SLOTS_START = ifThenElse((SlotTypeID==5), TARGET.RequestGPUs>0, ($(EXTRA_SLOTS_START)))
NEW_RESOURCES_LIST = $(NEW_RESOURCES_LIST) GPUs
...
```

In line:
```
SLOT_TYPE_5 = cpus=$(MACHINE_RESOURCE_GPUs), GPUs=$(MACHINE_RESOURCE_GPUs), ram=main
```
`ram` was being initialized to `main` which was incorrect. Looked back in the documentation to revisit the defining order of the parameters in the resource description. The order expected was
`name, number, memory, type, disk` and I realized that there was an additional comma required to appropriately indicate that `main` corresponds to `type` and not `ram`. My definition read:
```
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="ioslot,2,disk=1GB;monitor;GPUs,3,main"/>
```

Instead, it had to be (note the additional comma when defining GPUs in `value`):
```
<attr name="GLIDEIN_Resource_Slots" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="ioslot,2,disk=1GB;monitor;GPUs,3,,main"/>
```

Despite the change, the same message about percent or fraction of disk not being defined still persisted, which looked like below:
```
...
11/04/24 09:48:02 (pid:850266) Startd will not enforce disk limits via logical volume management.
11/04/24 09:48:02 (pid:850266) Local machine resource GPUs = 0
11/04/24 09:48:02 (pid:850266) Local machine resource ioslot = 2
11/04/24 09:48:02 (pid:850266) Local machine resource monitor = 1
11/04/24 09:48:02 (pid:850266) Got execute_dir = /var/lib/condor/execute/dir_827672/glide_t7KDtF/execute
11/04/24 09:48:02 (pid:850266) Got execute_dir = /var/lib/condor/execute/dir_827672/glide_t7KDtF/execute
11/04/24 09:48:02 (pid:850266) You must specify a percent or fraction for disk in slot type 3
```

As a next step, I removed `GPUs` declaration from `value` and tried submitting jobs and the issue continued to persist. Further, I removed `monitor` declaration from `value` and again tried submitting jobs, but the issue did not get resolved. Even though I was using an example that was described in the documentation, turns out the message about having to specify percentage or fraction was occurring due to disk not being defined for `ioslot` resource in `value` as per the error message. 

What I had was `value="ioslot,2,disk=1GB"` which was from the documentation. But going as per the error message, I changed it to say `value="ioslot,2,disk=25%"` which resolved the problem and the glidein started successfully and completed the submitted jobs too.

As an extension, I updated the `GLIDEIN_Resource_Slots` `attr` tag’s value from `value="ioslot,2,disk=1GB;monitor;GPUs,3,,main"` to value="ioslot,2,disk=10%;monitor;GPUs,3,,main" . The glidein startup completed without any errors (including condor startup) and the submitted jobs were also completed, **therefore passing the testcase (TC7) that had previously failed.**

namrathaurs commented:
> Something I was not sure of is why this particular test case expects disk space to be explicitly defined while other test cases did not report having to set disk space explicitly. Is this a side effect of defining three different types of resources as opposed to only one or two? Or, could something else be at play here (maybe a bug?)

When coming up with various test cases, the `value` definitions were selected/adapted based on the examples outlined in the documentation for `GLIDEIN_Resource_Slots` (under Custom Variables). There was only one example where disk space was explicitly defined and when defined, storage measurement unit (such as GB) was used. However, the description for `disk` category, in the documentation, said: `Disk space must be a fraction or percentage: 1/4, 25%. auto lets HTCondor do the splitting.` Further, description for `memory` category said: `Characters may be appended to a numerical value of memory to indicate percentage (%) or units. ...` It looks like the setting `disk=1GB` might have creeped in to the documentation as a result of an unintentional typo when that section was last written/updated.
